TechAndCheck/zenodotus
# ActiveJob queue tracker

This PR moves all scraping tasks to `ActiveJob` and creates a status page to view the `ActiveJob`/sidekiq queue. The status page is updated using `ActionCable`, which pushes job queue changes to a websocket connection that the front end of Zenodotus subscribes to. 

Beginning with this PR, we'll need redis active to run Zenodotus in dev, as Sidekiq uses it for storage. 

To test:
1. `bundle install` (to install Sidekiq)
2. Run a local redis server
3. `bundle exec sidekiq`
4. `rails t`. Note: rails tests won't engage Sidekiq. To do that, run the following in the rails console:
```ruby
require 'sidekiq/api'
tweet_urls = [
  "https://twitter.com/JasonKander/status/1478016167679864836",
  "https://twitter.com/smwat/status/1478022399555739652",
  "https://twitter.com/greg_doucette/status/1477696042535735296",
  "https://twitter.com/BillAdairDuke/status/1478022935587614723",
  "https://twitter.com/seungminkim/status/1478022908538720261",
  "https://twitter.com/normative/status/1478021120251023361",
  "https://twitter.com/jeffjarvis/status/1478019870478548993",
  "https://twitter.com/thehill/status/1477678824280182790",
  "https://twitter.com/tressiemcphd/status/1476978397679075336",
  "https://twitter.com/RepLizCheney/status/1477723817200041988",
  "https://twitter.com/tressiemcphd/status/1478009292926201856",
  "https://twitter.com/ShaneGoldmacher/status/1477997873010819078",
  "https://twitter.com/farhip/status/1478000944042414082",
  "https://twitter.com/JonathanCohn/status/1477987859005116420",
  "https://twitter.com/PunchbowlNews/status/1478006865296601091",
  "https://twitter.com/NadiehBremer/status/1478003463430811650",
  "https://twitter.com/Martina/status/1477741805802246149",
  "https://twitter.com/TheEconomist/status/1478000715436072960",
  "https://twitter.com/brianstelter/status/1478000387185692673",
  "https://twitter.com/tcwittes/status/1477998719207456770",
  "https://twitter.com/frostyhk/status/1477872373072535552",
  "https://twitter.com/shmcminn/status/1477994333802541061",
  "https://twitter.com/ddale8/status/1477993680271253507",
  "https://twitter.com/NPRKelly/status/1477993406722985984",
  ]

tweet_urls.each do |url|
	Sources::Tweet.create_from_url! url
	puts "queue size: #{Sidekiq::Queue.new.size}"
end
```

You should see logs of the queue size, showing that the scraper tasks are being handled by SideKiq. Additionally, if you watch http://localhost:3000/jobs, you should see ActionCable updating the jobs tracker table there.  
To do list before merging this PR:
- [ ] Figure out how to manage Sidekiq workers' simultaneous requests to `Zorki`/`Forki`. Add more instances of the scrapers? Make them multithreaded?
- [ ]  Shrink the diff on this PR by merging #105 to master (which requires a little work on #105, namely adding some `erb` files)
